src/formats/SMI.cpp

```diff
@@ -0,0 +1,705 @@
+static bool all(const std::vector<bool>& vec);
+
+SMIFormat::SMIFormat(const std::string& path, File::Mode mode, File::Compression compression)
+    : file_(TextFile::open(std::move(path), mode, compression))
+{
+    while (!file_->eof()) {
+        auto position = file_->tellg();
+        if (!file_ || position == std::streampos(-1)) {
+            throw format_error("IO error while reading '{}' as SMI", path);
+        }
+        if (forward(*file_)) {
+            steps_positions_.push_back(position);
+        }
+    }
+    file_->rewind();
+}
+
+size_t SMIFormat::nsteps() {
+    return steps_positions_.size() - 1;
+}
+
+void SMIFormat::read_step(const size_t step, Frame& frame) {
+    assert(step < steps_positions_.size());
+    file_->seekg(steps_positions_[step]);
+    read(frame);
+}
+
+/// Attempts to read a number from the string. The `start` argument is updated
+/// to the last numberic character in `smiles`.
+static size_t read_number(const std::string& smiles, size_t& start) {
+    size_t number = 0;
+    if (start >= smiles.size() || std::isdigit(smiles[start]) == 0) {
+        --start;
+        return number;
+    }
+    do {
+        number *= 10;
+        number += static_cast<size_t>(smiles[start] - '0');
+        ++start;
+    } while (start < smiles.size() && (std::isdigit(smiles[start]) != 0));
+    --start;
+    return number;
+}
+
+static std::set<char> aromatic_organic {'b', 'c', 'n', 'o', 's', 'p'};
+
+// Note: Including H here is non-standard, but allows for HC(H)(H)H
+static std::set<std::string> aliphatic_organic {"B", "C", "N", "O", "S", "P",
+                                                "F", "Cl", "Br", "I", "H"};
+
+// Valid chirality tags
+static std::set<std::string> chirality_tags {"TH", "SP", "TB", "OH", "AL"};
+
+static std::string read_property_atom(const std::string& smiles, size_t& i) {
+    auto old = i;
+    if (smiles[i] == '\'') {
+        do {
+             ++i;
+        } while( i < smiles.size() && smiles[i] != '\'');
+        ++i;
+        return smiles.substr(old + 1, i - old - 2);
+    } else {
+        do {
+            ++i;
+        } while( i < smiles.size() && std::islower(smiles[i]) != 0);
+        return smiles.substr(old, i - old);
+    }
+}
+
+Atom& SMIFormat::add_atom(Topology& topol, const std::string& atom_name) {
+    topol.add_atom(Atom(atom_name));
+
+    if (!first_atom_) {
+        topol.add_bond(previous_atom_, ++current_atom_, current_bond_order_);
+    }
+
+    first_atom_ = false;
+    previous_atom_ = current_atom_;
+    current_bond_order_ = Bond::SINGLE;
+    mol_vector_.back().add_atom(topol.size() - 1);
+    return topol[topol.size() - 1];
+}
+
+void SMIFormat::process_property_list_(Topology& topol, const std::string& smiles, size_t& i) {
+    double mass = 0.0;
+    if (std::isdigit(smiles[i]) != 0) {
+        mass = static_cast<double>(read_number(smiles, i));
+        ++i;
+    }
+
+    bool is_aromatic = std::islower(smiles[i]) != 0;
```

Comment:
std::islower uses the current user locale, which may break the code here for non-ascii atom names. We may want to write our own. There are already uses of such functions from cctype in the code, so this can be another PR.

Reply:
Hmmm. SMILES strings shouldn't have non-ascii atom names! Definitely a separate PR.